### PR TITLE
perf: use `vec_map` paging mmu for executor memory

### PR DESCRIPTION
**Pull request by tqn**
Some areas to potentially improve:
- `memory_checkpoint` always allocates as much space (for page pointers) as the largest key (page address). This is not necessary. Perhaps it should be a `VecDeque` of `Page`s along with its least address, or perhaps a `HashMap` or `BTreeMap` would do better. This choice probably depends on the shard size and memory access patterns.
- `memory_checkpoint` is `PagedMemory<Option<MemoryRecord>>` which is internally stored as `Vec<Option<Option<MemoryRecord>>>`. This is not the best for space.
- Loads of constant arithmetic at the top of `memory.rs`. Could be made into `const` functions taking (the size of) the generic value `V`.

### CODE REVIEW

File: crates/core/executor/src/memory.rs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    fn default() -> Self {
+        Self(VecMap::default())
+    }
+}
+
+/// Paged memory. Balances both memory locality and total memory usage.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PagedMemory<V> {
+    /// The internal page table.
+    pub page_table: VecMap<Page<V>>,
```
Comment (ctian1) [inline crates/core/executor/src/memory.rs]: VecMap actually has [worst case O(n) lookup and performance](https://docs.rs/vecmap-rs/latest/vecmap/), it seems incorrect to use in both places here?

File: crates/core/executor/src/state.rs [hunk 2]
```diff
@@ -40,14 +39,10 @@ pub struct ExecutionState {
 
     /// The memory which instructions operate over. Values contain the memory value and last shard
     /// + timestamp that each memory address was accessed.
-    #[serde(
-        serialize_with = "serialize_hashmap_as_vec",
-        deserialize_with = "deserialize_hashmap_as_vec"
-    )]
-    pub memory: HashMap<u32, MemoryRecord, BuildNoHashHasher<u32>>,
+    pub memory: PagedMemory<MemoryRecord>,
```
[the remark this answers is not in the document]
Comment (ctian1) [reply crates/core/executor/src/state.rs:42]: nvm, VecMap uses Option

File: crates/core/executor/src/memory.rs (hunk 1 above)
Comment (tqn) [reply crates/core/executor/src/memory.rs]: This is the crate `vec_map`.

File: crates/core/executor/src/memory.rs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+use serde::{Deserialize, Serialize};
+use vec_map::VecMap;
+
+use crate::events::MemoryRecord;
+
+/// The base 2 logarithm of the (maximum) page size in bytes.
+pub const LOG_PAGE_SIZE: usize = 12;
+/// The base 2 logarithm of the length of each page, considered as an array of `Option<MemoryRecord>`.
+pub const LOG_PAGE_LEN: usize =
+    LOG_PAGE_SIZE - size_of::<Option<MemoryRecord>>().next_power_of_two().ilog2() as usize;
```
[the remark this answers is not in the document]
Comment (tqn) [reply crates/core/executor/src/memory.rs:12]: I commented on this above in the PR. A couple concerns I had:
- Writing a bunch of `const` functions is extra work.
- Rust has many restrictions on const generics, e.g. type generic parameters cannot be used in const generic positions. By genericizing all the constants, we may restrict ourselves. This should not currently be an issue, but it is an issue if, say, we wanted to replace the `Page` `VecMap`s with arrays.

I feel that it's best to use logarithm when dealing with powers of two. This enables the use of the shift operators, which can be more efficient than dividing. I don't believe the compiler will "optimize" division/multiplication by powers of two because of truncation/overflow issues (at least for multiplication).

For that specific case in `indices`, it may be more efficient to do division and remainder together (since x86 uses a single instruction to do both), rather than shifting and masking. However, these sort of bitwise operations seem to be rather common in performance-critical code, which is why I use them.

In summary, I don't know which approach is more efficient. I would likely have to write micro-benchmarks to detect differences.

[the remark this answers is not in the document]
Comment (tqn) [reply crates/core/executor/src/memory.rs]: Moved all the constants into the `impl` block.

**Review by ctian1 — APPROVED** [review]
